Lib/test/test_import/__init__.py

```diff
@@ -689,6 +689,27 @@ def test_absolute_import_without_future(self):
             self.fail("explicit relative import triggered an "
                       "implicit absolute import")
 
+    def test_import_from_non_package(self):
+        sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'data', 'package2'))
+        try:
+            with self.assertRaises(ImportError):
+                import submodule1
+            self.assertNotIn('submodule1', sys.modules)
+            self.assertNotIn('submodule2', sys.modules)
+        finally:
+            del sys.path[0]
+
+    def test_import_from_unloaded_package(self):
+        sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'data'))
+        try:
+            import package2.submodule1
+            package2.submodule1.submodule2
+        finally:
```

Comment:
All of this cleanup can be handled by the `test.test_importlib.util.import_state` and `uncache` context managers.